README.md
# go-avauth-oauth2

This is a oauth2 provider that use the aventer backend systems to authenticate. To be honest, it makes no sense for other peoples to use it, but I think it's a interesting starting point for the own oauth2 project. So, feel free and have a look around.

## How to use

This OAuth provider need the following information as environment variable.

```bash
export AUTH_SERVER=https://<AUTH SERVER>
export CLIENTDOMAIN=
export CLIENTID=
export CLIENTSECRET=
export GROUP=<USER GROUP>
export IDENTIFIER=
export JWT_SIGNKEY=<A VERY STRONG AND LONG PHRASE>
export LOGLEVEL=<debug|warn|info>
export REDIS_SERVER=<REDIS_SERVER>:6379

go run main.go

```




changelog.md
## master

- update go modules
- migrate to oauth v4


main.go
package main

import (
	"net/http"
	"os"

	"github.com/AVENTER-UG/util"
	"github.com/go-redis/redis"
	"github.com/go-session/session"
	"github.com/sirupsen/logrus"
	oredis "gopkg.in/go-oauth2/redis.v4"
	"gopkg.in/oauth2.v4/manage"
	"gopkg.in/oauth2.v4/models"
	"gopkg.in/oauth2.v4/server"
	"gopkg.in/oauth2.v4/store"
)

var (
	globalSessions *session.Manager
	JwtSignKey     string
	AuthServer     string
	ClientDomain   string
	ClientID       string
	ClientSecret   string
	LogLevel       string
	AppName        string
	EnableSyslog   bool
	UserGroup      string
	Identifier     string
	RedisServer    string
	RedisDB        string
	Cache          *redis.Client
	Srv            *server.Server
)

func main() {
	AuthServer = os.Getenv("AUTH_SERVER")
	ClientDomain = os.Getenv("CLIENTDOMAIN")
	ClientID = os.Getenv("CLIENTID")
	ClientSecret = os.Getenv("CLIENTSECRET")
	LogLevel = os.Getenv("LOGLEVEL")
	UserGroup = os.Getenv("GROUP")
	RedisServer = os.Getenv("REDIS_SERVER")
	JwtSignKey = os.Getenv("JWT_SIGNKEY")

	util.SetLogging(LogLevel, EnableSyslog, AppName)

	logrus.Info("ISPCONFIGServer=", AuthServer)
	logrus.Info("ClientDomain=", ClientDomain)
	logrus.Info("ClientSecret=", ClientSecret)
	logrus.Info("ClientID=", ClientID)

	initCache()

	manager := manage.NewDefaultManager()
	manager.SetAuthorizeCodeTokenCfg(manage.DefaultAuthorizeCodeTokenCfg)

	// token store
	manager.MapTokenStorage(oredis.NewRedisStore(&redis.Options{
		Addr: RedisServer,
	}))

	// generate jwt access token
	manager.MapAccessGenerate(&JWTGenerator{SignedKey: []byte(JwtSignKey)})

	clientStore := store.NewClientStore()
	clientStore.Set(ClientID, &models.Client{
		ID:     ClientID,
		Secret: ClientSecret,
		Domain: ClientDomain,
	})
	manager.MapClientStorage(clientStore)

	Srv = server.NewServer(server.NewConfig(), manager)

	Srv.SetPasswordAuthorizationHandler(srvPasswordAuthorizationHandler)
	Srv.SetUserAuthorizationHandler(userAuthorizeHandler)
	Srv.SetInternalErrorHandler(srvInternalErrorHandler)
	Srv.SetResponseErrorHandler(srvResponseErrorHandler)

	http.HandleFunc("/oauth/login", loginHandler)
	http.HandleFunc("/oauth/auth", authHandler)
	http.HandleFunc("/oauth/userinfo", userInfoHandler)
	http.HandleFunc("/oauth/authorize", authorizeHandler)
	http.HandleFunc("/oauth/token", tokenHandler)
	http.HandleFunc("/health", apiHealth)

	logrus.Println("Server is running at 9096 port.")
	logrus.Fatal(http.ListenAndServe(":9096", nil))
}

func outputHTML(w http.ResponseWriter, req *http.Request, filename string) {
	file, err := os.Open(filename)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer file.Close()
	fi, _ := file.Stat()
	http.ServeContent(w, req, file.Name(), fi.ModTime(), file)
}

func sendJSON(js []byte, w http.ResponseWriter) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Write(js)
}

// init the redis cache
func initCache() {
	client := redis.NewClient(&redis.Options{
		Addr: RedisServer,
		DB:   1,
	})

	pong, err := client.Ping().Result()
	logrus.Debug(pong, err)

	Cache = client
}


token.go
package main

import (
	"context"

	jwt "github.com/golang-jwt/jwt"
	"github.com/sirupsen/logrus"
	"gopkg.in/oauth2.v4"

	cfg "github.com/AVENTER-UG/go-avauth-oauth2/types"
)

type JWTGenerator struct {
	SignedKey []byte
}

func (a *JWTGenerator) Token(ctx context.Context, data *oauth2.GenerateBasic, isGenRefresh bool) (access, refresh string, err error) {
	logrus.Debug("Token: LoggedIn: ", data)
	signingKey := []byte(JwtSignKey)

	// Create the Claims
	myClaims := &cfg.CustomClaims{
		ClientID: data.UserID,
		Type:     "user",
		StandardClaims: jwt.StandardClaims{
			Audience:  data.Client.GetID(),
			Subject:   data.UserID,
			ExpiresAt: data.TokenInfo.GetAccessCreateAt().Add(data.TokenInfo.GetAccessExpiresIn()).Unix(),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, myClaims)
	access, err = token.SignedString(signingKey)

	if err != nil {
		logrus.Debug("newToken = ", err)
	}

	return
}

// Check if the token is valid
// return true and false and if its true then also the clientID of the token
func validToken(token string) (bool, *cfg.CustomClaims) {
	newToken, err := jwt.ParseWithClaims(token, &cfg.CustomClaims{}, func(newToken *jwt.Token) (interface{}, error) {
		return []byte(JwtSignKey), nil
	})

	if err != nil {
		logrus.Error("validToken: ", err)
		return false, nil
	}

	if claims, ok := newToken.Claims.(*cfg.CustomClaims); ok && newToken.Valid {
		return true, claims
	} else {
		return false, nil
	}
}


avauth.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/sirupsen/logrus"

	cfg "github.com/AVENTER-UG/go-avauth-oauth2/types"
)

// AuthUser Function to authenticate a user against the ispconfig backend system
//	curl -X GET -u <customeruser>:<password>  http://127.0.0.1:10777/api/v0/authUser
//

func AuthUser(username, password string) cfg.CheckAuth {
	// check user auth
	client := &http.Client{}
	req, err := http.NewRequest("GET", AuthServer+"/checkAuth.php", nil)

	var user cfg.CheckAuth

	// User authorization not correct
	if err != nil {
		logrus.Error("authUser 1: ", err)
		user.Auth = false
		return user
	}
	req.SetBasicAuth(username, password)
	res, err := client.Do(req)

	if err != nil {
		logrus.Error("authUser 2: ", err)
		user.Auth = false
	} else {
		json.NewDecoder(res.Body).Decode(&user)
	}

	logrus.Debug("AuthUser: user: ", user)

	ug := strings.Split(user.Client.GroupAdditional, ":")

	// if the server was started with UserGroup then check if the user in in that group
	if UserGroup != "" {
		if len(ug) >= 2 {
			if !strings.Contains(ug[1], UserGroup) {
				user.Auth = false
			}
		} else {
			user.Auth = false
		}
	}

	logrus.Debug("AuthUser: ug: ", ug, " UserGroup: ", UserGroup)

	return user
}


health.go
package main

import (
	"net/http"

	"github.com/sirupsen/logrus"
)

func apiHealth(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	client := &http.Client{}

	logrus.Debug("Health Check")

	// check auth server connection
	req, _ := http.NewRequest("GET", AuthServer+"/status.php", nil)
	res, err := client.Do(req)

	if err != nil {
		logrus.Error("Health to Auth Server: ", err)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	if res.StatusCode != 200 {
		logrus.Error("Health to Auth Server: ", res.StatusCode)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	_, err = Cache.Ping().Result()

	if err != nil {
		logrus.Error("Health to Redis Server: ", err)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
}


handler.go
package main

import (
	"encoding/json"
	"net/http"
	"net/url"

	"github.com/go-session/session"
	"github.com/sirupsen/logrus"
)

func tokenHandler(w http.ResponseWriter, r *http.Request) {
	logrus.Debug("/token")
	err := Srv.HandleTokenRequest(w, r)
	if err != nil {
		logrus.Error("/token: Error ", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
}

func authorizeHandler(w http.ResponseWriter, r *http.Request) {
	logrus.Debug("/authorize")
	store, err := session.Start(nil, w, r)
	if err != nil {
		logrus.Error("/authorize: Error ", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	var form url.Values
	if v, ok := store.Get("ReturnUri"); ok {
		logrus.Debug("/authorize: ReturnUri: ", v, ok)
		form = v.(url.Values)
	}
	r.Form = form

	store.Delete("ReturnUri")
	store.Save()

	err = Srv.HandleAuthorizeRequest(w, r)
	if err != nil {
		logrus.Error("/authorize: Error1 ", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
	}
}

func userInfoHandler(w http.ResponseWriter, r *http.Request) {
	logrus.Debug("/userinfo")
	token, err := Srv.ValidationBearerToken(r)
	if err != nil {
		logrus.Error("/userinfo: Error ", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	//var info []byte

	info, err := json.Marshal(token)

	if err != nil {
		logrus.Error("userInfoHandler: Error Create JSON")
		return
	}

	sendJSON(info, w)
}

func userAuthorizeHandler(w http.ResponseWriter, r *http.Request) (userID string, err error) {
	logrus.Debug("userAuthorizeHandler")
	store, err := session.Start(nil, w, r)
	if err != nil {
		logrus.Error("userAuthorizeHandler: Error ", err)
		return
	}

	uid, ok := store.Get("LoggedInUserID")
	if !ok {
		logrus.Debug("userAuthorizeHandler: Not LoggedIn")
		if r.Form == nil {
			r.ParseForm()
		}

		store.Set("ReturnUri", r.Form)
		store.Save()

		w.Header().Set("Location", "/oauth/login")
		w.WriteHeader(http.StatusFound)
		return
	}

	logrus.Debug("userAuthorizeHandler: LoggedIn: ", uid)
	userID = uid.(string)
	store.Delete("LoggedInUserID")
	store.Save()
	return
}

func loginHandler(w http.ResponseWriter, r *http.Request) {
	logrus.Debug("loginHandler")
	store, err := session.Start(nil, w, r)
	if err != nil {
		logrus.Error("loginHandler: Error ", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if r.Method == "POST" {

		username := r.FormValue("username")
		userPassword := r.FormValue("password")
		ispUser := AuthUser(username, userPassword)
		logrus.Debug("loginHandler: ispUser: ", ispUser)

		if ispUser.Auth == true {

			logrus.Debug("loginHandler: Auth True")
			store.Set("LoggedInUserID", ispUser.Client.ClientID)

			store.Save()

			w.Header().Set("Location", "/oauth/auth")
			w.WriteHeader(http.StatusFound)
			return

		}

	}
	outputHTML(w, r, "static/login.html")
}

func authHandler(w http.ResponseWriter, r *http.Request) {
	logrus.Debug("authHandler")
	store, err := session.Start(nil, w, r)
	if err != nil {
		logrus.Error("authHandler: Error ", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if _, ok := store.Get("LoggedInUserID"); !ok {
		logrus.Debug("authHandler: Not LoggedIn")
		w.Header().Set("Location", "/oauth/login")
		w.WriteHeader(http.StatusFound)
		return
	}

	outputHTML(w, r, "static/auth.html")
}


types/types.go
package types

import (
	"github.com/golang-jwt/jwt"
)

type CheckAuth struct {
	Method string `json:"method"`
	Client struct {
		Username        string `json:"username"`
		EMail           string `json:"email"`
		Type            string `json:"type"`
		ClientID        string `json:"client_id"`
		Language        string `json:"language"`
		Country         string `json:"country"`
		GroupMaster     string `json:"group_master"`
		GroupAdditional string `json:"group_additional"`
		Notes           string `json:"notes"`
	} `json:"client"`
	Token string `json:"token"`
	Auth  bool   `json:"auth"`
}

type CustomClaims struct {
	Type            string `json:"type"`
	ClientID        string `json:"client_id"`
	GroupMaster     string `json:"group_master"`
	GroupAdditional string `json:"group_additional"`
	Notes           string `json:"notes"`
	jwt.StandardClaims
}


server.go
package main

import (
	"encoding/json"
	"net/http"
	"net/url"
	"os"

	"git.aventer.biz/AVENTER/util"
	"github.com/go-redis/redis"
	"github.com/go-session/session"
	"github.com/sirupsen/logrus"
	oredis "gopkg.in/go-oauth2/redis.v3"
	"gopkg.in/oauth2.v3/errors"
	"gopkg.in/oauth2.v3/manage"
	"gopkg.in/oauth2.v3/models"
	"gopkg.in/oauth2.v3/server"
	"gopkg.in/oauth2.v3/store"
)

var (
	globalSessions *session.Manager
	JwtSignKey     string
	AuthServer     string
	ClientDomain   string
	ClientID       string
	ClientSecret   string
	User           UserInfo
	LogLevel       string
	AppName        string
	EnableSyslog   bool
	UserGroup      string
	Identifier     string
	RedisServer    string
	RedisDB        string
)

type UserInfo struct {
	UserID      string `json:"user_id"`
	UserName    string `json:"user_name"`
	EMail       string `json:"email"`
	ConnectorID string `json:"connector_id"`
	Sub         string `json:"sub"`
	Name        string `json:"name"`
	Auth        string `json:"auth"`
	ClientID    string `json:"client_id"`
	Type        string `json:"type"`
}

func main() {
	AuthServer = os.Getenv("AUTH_SERVER")
	ClientDomain = os.Getenv("CLIENTDOMAIN")
	ClientID = os.Getenv("CLIENTID")
	ClientSecret = os.Getenv("CLIENTSECRET")
	LogLevel = os.Getenv("LOGLEVEL")
	UserGroup = os.Getenv("GROUP")
	RedisServer = os.Getenv("REDIS_SERVER")
	JwtSignKey = os.Getenv("JWT_SIGNKEY")

	util.SetLogging(LogLevel, EnableSyslog, AppName)

	logrus.Info("ISPCONFIGServer=", AuthServer)
	logrus.Info("ClientDomain=", ClientDomain)
	logrus.Info("ClientSecret=", ClientSecret)
	logrus.Info("ClientID=", ClientID)

	manager := manage.NewDefaultManager()
	manager.SetAuthorizeCodeTokenCfg(manage.DefaultAuthorizeCodeTokenCfg)

	// token store
	manager.MapTokenStorage(oredis.NewRedisStore(&redis.Options{
		Addr: RedisServer,
	}))

	// generate jwt access token
	manager.MapAccessGenerate(&JWTGenerator{SignedKey: []byte(JwtSignKey)})
	//	manager.MapAccessGenerate(generates.NewAccessGenerate())

	clientStore := store.NewClientStore()
	clientStore.Set(ClientID, &models.Client{
		ID:     ClientID,
		Secret: ClientSecret,
		Domain: ClientDomain,
	})
	manager.MapClientStorage(clientStore)

	srv := server.NewServer(server.NewConfig(), manager)

	srv.SetPasswordAuthorizationHandler(func(username, password string) (userID string, err error) {
		ispUser := AuthUser(username, password)
		logrus.Debug("SetPasswordAuthorizationHandler:", ispUser)

		if ispUser.Auth == false {
			logrus.Debug("SetPasswordAuthorizationHandler: Auth False")
			return "", err
		}

		return ispUser.Client.ClientID, nil
	})

	srv.SetUserAuthorizationHandler(userAuthorizeHandler)

	srv.SetInternalErrorHandler(func(err error) (re *errors.Response) {
		logrus.Error("SetInternalErrorHandler: Internal Error:", err.Error())
		return
	})

	srv.SetResponseErrorHandler(func(re *errors.Response) {
		logrus.Error("SetResponseErrorHandler: Response Error:", re.Error.Error())
	})

	http.HandleFunc("/oauth/login", loginHandler)
	http.HandleFunc("/oauth/auth", authHandler)
	http.HandleFunc("/oauth/userinfo", func(w http.ResponseWriter, r *http.Request) {
		logrus.Debug("/userinfo")
		token, err := srv.ValidationBearerToken(r)
		if err != nil {
			logrus.Error("/userinfo: Error ", err)
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		//var info []byte
		info, err := json.Marshal(token)

		if err != nil {
			logrus.Error("userInfoHandler: Error Create JSON")
			return
		}

		sendJSON(info, w)
	})

	http.HandleFunc("/oauth/authorize", func(w http.ResponseWriter, r *http.Request) {
		logrus.Debug("/authorize")
		store, err := session.Start(nil, w, r)
		if err != nil {
			logrus.Error("/authorize: Error ", err)
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		var form url.Values
		if v, ok := store.Get("ReturnUri"); ok {
			logrus.Debug("/authorize: ReturnUri: ", v, ok)
			form = v.(url.Values)
		}
		r.Form = form

		store.Delete("ReturnUri")
		store.Save()

		err = srv.HandleAuthorizeRequest(w, r)
		if err != nil {
			logrus.Error("/authorize: Error1 ", err)
			http.Error(w, err.Error(), http.StatusBadRequest)
		}
	})

	http.HandleFunc("/oauth/token", func(w http.ResponseWriter, r *http.Request) {
		logrus.Debug("/token")
		err := srv.HandleTokenRequest(w, r)
		if err != nil {
			logrus.Error("/token: Error ", err)
			http.Error(w, err.Error(), http.StatusInternalServerError)
		}
	})

	logrus.Println("Server is running at 9096 port.")
	logrus.Fatal(http.ListenAndServe(":9096", nil))
}

func userAuthorizeHandler(w http.ResponseWriter, r *http.Request) (userID string, err error) {
	logrus.Debug("userAuthorizeHandler")
	store, err := session.Start(nil, w, r)
	if err != nil {
		logrus.Error("userAuthorizeHandler: Error ", err)
		return
	}

	uid, ok := store.Get("LoggedInUserID")
	if !ok {
		logrus.Debug("userAuthorizeHandler: Not LoggedIn")
		if r.Form == nil {
			r.ParseForm()
		}

		store.Set("ReturnUri", r.Form)
		store.Save()

		w.Header().Set("Location", "/oauth/login")
		w.WriteHeader(http.StatusFound)
		return
	}

	logrus.Debug("userAuthorizeHandler: LoggedIn: ", uid)
	userID = uid.(string)
	store.Delete("LoggedInUserID")
	store.Save()
	return
}

func loginHandler(w http.ResponseWriter, r *http.Request) {
	logrus.Debug("loginHandler")
	store, err := session.Start(nil, w, r)
	if err != nil {
		logrus.Error("loginHandler: Error ", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if r.Method == "POST" {

		username := r.FormValue("username")
		userPassword := r.FormValue("password")
		ispUser := AuthUser(username, userPassword)
		logrus.Debug("loginHandler: ispUser: ", ispUser)

		if ispUser.Auth == true {

			logrus.Debug("loginHandler: Auth True")
			store.Set("LoggedInUserID", ispUser.Client.ClientID)

			store.Save()

			w.Header().Set("Location", "/oauth/auth")
			w.WriteHeader(http.StatusFound)
			return

		}

	}
	outputHTML(w, r, "static/login.html")
}

func authHandler(w http.ResponseWriter, r *http.Request) {
	logrus.Debug("authHandler")
	store, err := session.Start(nil, w, r)
	if err != nil {
		logrus.Error("authHandler: Error ", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if _, ok := store.Get("LoggedInUserID"); !ok {
		logrus.Debug("authHandler: Not LoggedIn")
		w.Header().Set("Location", "/oauth/login")
		w.WriteHeader(http.StatusFound)
		return
	}

	outputHTML(w, r, "static/auth.html")
}

func outputHTML(w http.ResponseWriter, req *http.Request, filename string) {
	file, err := os.Open(filename)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer file.Close()
	fi, _ := file.Stat()
	http.ServeContent(w, req, file.Name(), fi.ModTime(), file)
}

func sendJSON(js []byte, w http.ResponseWriter) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Write(js)
}
